### ISSUE TITLE
'podman pull' no simple way to check if newer image was pulled

### ISSUE THREAD

**Issue by andrin55** [containers/podman#13582]
/kind feature

**Description**

It would be nice if there was a status line when pulling an image to determine if a newer version was pulled. This would help with automation.

**Describe the results you received:**
Example with podman pulling an image which is new:
```
podman pull docker.io/busybox
Trying to pull docker.io/library/busybox:latest...
Getting image source signatures
Copying blob 554879bb3004 done  
Copying config 2fb6fc2d97 done  
Writing manifest to image destination
Storing signatures
2fb6fc2d97e10c79983aa10e013824cc7fc8bae50630e32159821197dda95fe3
```

Example with podman pulling the same image again:
```
podman pull docker.io/busybox
Trying to pull docker.io/library/busybox:latest...
Getting image source signatures
Copying blob 554879bb3004 skipped: already exists  
Copying config 2fb6fc2d97 done  
Writing manifest to image destination
Storing signatures
2fb6fc2d97e10c79983aa10e013824cc7fc8bae50630e32159821197dda95fe3
```

One could only determine the image already was up-to-date by looking at the "Copying blob" lines.

**Describe the results you expected:**
Example with docker pull with a new image:
```
docker pull docker.io/busybox
Using default tag: latest
Trying to pull repository docker.io/library/busybox ...
latest: Pulling from docker.io/library/busybox
554879bb3004: Pull complete
Digest: sha256:caa382c432891547782ce7140fb3b7304613d3b0438834dce1cad68896ab110a
Status: Downloaded newer image for docker.io/busybox:latest
```

Example with docker pull on the same existing image:
```
docker pull docker.io/busybox
Using default tag: latest
Trying to pull repository docker.io/library/busybox ...
latest: Pulling from docker.io/library/busybox
Digest: sha256:caa382c432891547782ce7140fb3b7304613d3b0438834dce1cad68896ab110a
Status: Image is up to date for docker.io/busybox:latest
```

With docker pulling the image, it is very easy to distinguish between a new image "Downloaded newer image" and an unmodified one "Image is up to date".

**Output of `podman version`:**

```
Version:      3.4.2
API Version:  3.4.2
Go Version:   go1.16.12
Built:        Wed Feb  2 00:59:28 2022
OS/Arch:      linux/amd64
```

**Comment by vrothberg** [comment]
Thanks for reaching out, @andrin55!

I think there's some valuable information in the Docker output that we could use as well.

@mtrmac WDYT?

**Comment by mtrmac** [comment]
(Note that this is _only_ affecting the human-readable log output; it would not show up in the API.)

**Comment by rhatdan** [comment]
@mtrmac @vrothberg Anything ever happen with this?

**Comment by mtrmac** [comment]
Given
> (Note that this is only affecting the human-readable log output; it would not show up in the API.)

and the _very small_ impact on actual run time, this seems rather low-priority to me. It’s worth trying / investigating but not something I expect to prioritize at least within the next month.

**Comment by rhatdan** [comment]
Any objections to changing copy.Options.OptimizeDestinationImageAlreadyExists to default to on, or customizable in containers/common, so that we can see if this is better?

**Comment by mtrmac** [comment]
> Any objections to changing copy.Options.OptimizeDestinationImageAlreadyExists to default to on

To the extent it can be called an objection, “read the code / test this to verify it works as expected first, don’t just blindly try”.

> or customizable in containers/common, so that we can see if this is better?

I don’t see why this would be a per-site option. Either it should be a global default, or an option to decide per use (maybe based on the repo in question, and how it is used on that system); I find it hard to imagine a situation where there is something specific to the _machine_ that would drive such a decision, so that it would make sense to configure in `/etc`. Also, it doesn’t make much sense to me to recommend `OptimizeDestinationImageAlreadyExists` as a way to help automation, while at the same time letting users configure the system to disable that behavior, silently breaking that automation.

---

Note that turning `OptimizeDestinationImageAlreadyExists` on would only change the text output to stdout; it would not provide an API-consumable value (and the text of the output to stdout is _not_ at all an API commitment — if it were, we could’t do the contemplated change in the first place).

Re-reading the original report, because the original motivation is
> This would help with automation.

this is something between “completely useless” and “users would depend on it but their automation is going to break and they are going to be unhappy”. (Also, nitpicking, the `OptimizeDestinationImageAlreadyExists` does not trigger on some code paths that were originally thought to be too hard or too rare; it’s not a reliable way to tell that the image has _not_ been updated.)

To actually implement this as requested, some variant of 
> Wouldn't it be safer to just run something like `podman inspect -f '{{.Id}}' $IMAGE` before and after the pull and compare if they are different?

(or possibly getting the image ID before pull, and the image ID that is the outcome of pull and returned by `copy.Image`) would be necessary, whether outside Podman or as a built-in Podman feature providing an API-exposed (CLI-automation-consumable?) value.

**Comment by rhatdan** [comment]
On 5/25/22 15:09, Miloslav Trmač wrote:
>
>     Any objections to changing
>     copy.Options.OptimizeDestinationImageAlreadyExists to default to on
>
> To the extent it can be called an objection, “read the code / test 
> this to verify it works as expected first, don’t just blindly try”.
>
If you feel the code is buggy, then that is reasonable to leave off.  
Currently there is no way for user to turn this flag on, so the code 
will never get improved.  I would have thought that this flag would 
optimize pulling of images, but if you don't believe it will ,then what 
is the purpose of the flag?
>
>     or customizable in containers/common, so that we can see if this
>     is better?
>
> I don’t see why this would be a per-site option. Either it should be a 
> global default, or an option to decide per use (maybe based on the 
> repo in question, and how it is used on that system); I find it hard 
> to imagine a situation where there is something specific to the 
> /machine/ that would drive such a decision, so that it would make 
> sense to configure in |/etc|. Also, it doesn’t make much sense to me 
> to recommend |OptimizeDestinationImageAlreadyExists| as a way to help 
> automation, while at the same time letting users configure the system 
> to disable that behavior, silently breaking that automation.
>
Find then should it be added a customizable field in registries.conf?  I 
am not necessarily looking to solve this issue, but more to the fact 
that a fancy new option was brought to my attention and I want to know 
if it actually Optimizes anything, and if it does, I would like allow 
users to turn it on, or potentially turn it on by default and then allow 
users who have issues with it, turn it off.

**Comment by mtrmac** [comment]
> If you feel the code is buggy, then that is reasonable to leave off.

I’d not _known_ to be buggy. It just wasn’t originally made with pulls in mind, so it might need a look in that situation. (That was just a minor pushback to the idea that trying the option is a way to tell whether it works correctly.)

> I would have thought that this flag would optimize pulling of images, but if you don't believe it will ,then what is the purpose of the flag?

It doesn’t “optimize pulling” to be always better; if that were possible, we would just make that the default behavior and not an option. It’s users’ choice of a trade-off to make pulls of already-present images faster, by making pulls of not-yet-present images slower. It is, probably (no data), usually the wrong trade-off — but for pulls, assuming local storage accesses are negligible, it might be worth it.


> Find then should it be added a customizable field in registries.conf?

No, it’s not a per-registry trade-off either. It’s a per-user-behavior + per-repo-management-policy trade-off. The same repo might benefit from that flag when used by a frequent puller, and not benefit from it when used by a rare puller, or by someone who deletes pulled images.

>  I am not necessarily looking to solve this issue, but more to the fact that a fancy new option was brought to my attention and I want to know if it actually Optimizes anything

https://github.com/containers/image/blob/89f33ce0559760ee07567e00bb6dc0753b2468b2/copy/copy.go#L163-L165

**Comment by rhatdan** [comment]
@vrothberg @mtrmac What should we do with this issue?

**Comment by mtrmac** [comment]
See the end of https://github.com/containers/podman/issues/13582#issuecomment-1137751255 for what I think makes sense doing. I’m not going to work on making that a Podman-integrated feature for the foreseeable future.

**Comment by vrothberg** [comment]
I think `podman pull` should receive a new `--policy` flag that does the same as `podman create --pull` (i.e., pull should support pull policies).  This way, user can do a `podman pull --policy=newer`.  If an image gets pulled, it's newer.  No need to parse a progress bar.

**Comment by rhatdan** [comment]
SGTM

**Comment by mtrmac** [comment]
@vrothberg At least the original report mentions automation; so _assuming_ that is an important constraint (which we haven’t had explicitly confirmed, to be fair) there should be some way for the code invoking Podman to tell whether the image was pulled… maybe an exit code? OTOH a pull “succeeding” with a non-zero exit code seems unexpected.

I’m also not sure that users who ask “was a new image pulled” are going to look for an option driving _when_ we decide to pull an image. That feels fairly indirect to me.

Ultimately the user can already do this via https://github.com/containers/podman/issues/13582#issuecomment-1074898678 and I’m not sure any new option is going to make this _significantly_ easier.

---

… actually, more than a year after filing the original report… I am very unclear on why this is desirable at all. I mean, either the consumer needs a fresh image, and the unavoidable cost is having to pull, or the consumer doesn’t need a fresh image, and then pulling is not worth the extra effort.

What _decision_ is going to be made based on whether a new image was pulled? Especially when that information is only available _after_ the fact, i.e. after the costly time etc. to pull an image was already pulled.

@andrin55 could you clarify what decision do you plan to make based on pulling a new image, please? And to be explicit, is it scripted or by a human visually inspecting text?

**Comment by andrin55** [comment]
@mtrmac the idea behind the additional output is to run "podman pull" in a shell script or ansible playbook and determine based on the returned output if the container needs to be recreated. 

Additionally it would provide a more user friendly way to inform the user that a new image was downloaded instead of comparing the hash manually.

**Comment by mtrmac** [comment]
@andrin55 Thanks, that use case definitely makes sense.

It seems to me it doesn’t really matter whether _this pull_ pulled a different image — the question to ask is whether an image (which exists locally? or should this be asked about an image which is still on the registry and not yet pulled?) differs from the image currently running in the container. E.g. suppose the `:latest` tag is moved to a newer version, and then, due to a critical problem, rolled back to a previous one. In that case, [assuming for some reason that a previous pull did not trigger a container restart, e.g. because a script was aborted] even if the pull of the rollback did fetch a “new” (i.e. a different) image, if it is rolled back exactly to the image already running in the container, the container should not be restarted.

**Comment by vrothberg** [comment]
It is very similar (identical?) to what `podman auto-update` does.  An image will pulled (and a container using it updated) if the image on the registry differs (i.e., has a different digest).

I still believe we need something like `podman pull --policy=newer` since `podman pull` will otherwise always pull the image.

**Comment by mtrmac** [comment]
@vrothberg What’s the difference between fetching metadata, and not pulling the image because the image is locally available, and between always pulling the image, but upon reading the metadata, detecting the layers and the config are all present locally, and nothing needs to be written to storage?

To me, that’s almost a distinction with no difference. Strictly speaking, the set of metadata and the set of HTTP request is a bit different, and the progress messages obviously differ, but the net end-user effect is the same. Conceptually, a “check + pull if newer” should never _need_ to be faster than “pull + optimize”. We aren’t quite there (e.g. we always fetch the config), but if did have that performance characteristic, why make the user think whether `--policy=newer` should be used?

**Comment by vrothberg** [comment]
> @vrothberg What’s the difference between fetching metadata, and not pulling the image because the image is locally available, and between always pulling the image, but upon reading the metadata, detecting the layers and the config are all present locally, and nothing needs to be written to storage?

I don't see a big difference.

> [..] why make the user think whether --policy=newer should be used?

I don't feel strongly.  A `--policy=newer` seems straight forward to script as _any_ output indicates that something got pulled.

**Comment by mailinglists35** [comment]
tl;dr what's a non-simple way to check if my cached image is the same or different from upstream image?

**Comment by rhatdan** [comment]
skopeo inspect of remote image and compare ID to local image?

**Comment by gvenzl** [comment]
I just ran into this issue as well and would be grateful if there was an easy way to verify if the cached images is different from upstream.

**Comment by yeroc** [comment]
Running into this myself as well.  Similar use case to the one mentioned where we're doing automation and need to know whether to restart the service based on whether or not a pull resulted in a new image.
